# Healthcare Claims Fraud Detection

## Data Preprocessing for Machine Learning

This notebook prepares the provider-level dataset for machine learning by performing feature selection, train-test splitting, feature scaling, handling class imbalance using SMOTE, and creation of final modeling datasets.

In [17]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from imblearn.over_sampling import SMOTE
import warnings
warnings.filterwarnings("ignore")
print("Libraries Imported Successfully")

Libraries Imported Successfully


## Load Provider-Level Dataset

In [18]:
provider_df = pd.read_csv("../data/processed/provider_features.csv")
print("Dataset Loaded Successfully")

Dataset Loaded Successfully


## Dataset Overview

In [19]:
print("Shape:",provider_df.shape)
provider_df.info()

Shape: (5410, 20)
<class 'pandas.DataFrame'>
RangeIndex: 5410 entries, 0 to 5409
Data columns (total 20 columns):
 #   Column                         Non-Null Count  Dtype  
---  ------                         --------------  -----  
 0   Provider                       5410 non-null   str    
 1   PotentialFraud                 5410 non-null   str    
 2   InpatientClaimCount            5410 non-null   float64
 3   UniqueInpatientBeneficiaries   5410 non-null   float64
 4   AvgInpatientReimbursement      5410 non-null   float64
 5   TotalInpatientReimbursement    5410 non-null   float64
 6   AvgLengthOfStay                5410 non-null   float64
 7   MaxLengthOfStay                5410 non-null   float64
 8   AvgDiagnosisCount              5410 non-null   float64
 9   OutpatientClaimCount           5410 non-null   float64
 10  UniqueOutpatientBeneficiaries  5410 non-null   float64
 11  AvgOutpatientReimbursement     5410 non-null   float64
 12  TotalOutpatientReimbursement   5410 non-n

### Interpretation

All variables are complete with no missing values.

The dataset contains numerical features suitable for machine learning and a binary target variable representing provider fraud status.

## Feature Selection

In [20]:
features_to_drop = ["Provider", "PotentialFraud", "Fraud"]
X = provider_df.drop( columns=features_to_drop)
y = provider_df["Fraud"]
print("Feature Matrix Shape:", X.shape)
print("Target Vector Shape:", y.shape)

Feature Matrix Shape: (5410, 17)
Target Vector Shape: (5410,)


A total of 17 predictor variables were selected for modeling.

Identifier columns and the original target label were removed to prevent data leakage while retaining all engineered provider-level features.

## Review Modeling Features

In [21]:
X.columns.tolist()

['InpatientClaimCount',
 'UniqueInpatientBeneficiaries',
 'AvgInpatientReimbursement',
 'TotalInpatientReimbursement',
 'AvgLengthOfStay',
 'MaxLengthOfStay',
 'AvgDiagnosisCount',
 'OutpatientClaimCount',
 'UniqueOutpatientBeneficiaries',
 'AvgOutpatientReimbursement',
 'TotalOutpatientReimbursement',
 'AvgPatientAge',
 'AvgChronicConditions',
 'DeceasedRate',
 'AvgIPAnnualReimbursement',
 'AvgOPAnnualReimbursement',
 'UniqueBeneficiaries']

### Interpretation

The selected features include provider utilization metrics, reimbursement measures, beneficiary characteristics, and healthcare activity indicators.

Several features identified during exploratory analysis as strong fraud indicators are included in the modeling dataset.

## Split Dataset into Training and Testing Sets

In [22]:
X_train, X_test, y_train, y_test = train_test_split(
    X,
    y,
    test_size=0.20,
    random_state=42,
    stratify=y
)
print("X_train:", X_train.shape)
print("X_test :", X_test.shape)
print("y_train:", y_train.shape)
print("y_test :", y_test.shape)

X_train: (4328, 17)
X_test : (1082, 17)
y_train: (4328,)
y_test : (1082,)


### Interpretation

The dataset was successfully divided into training and testing subsets using an 80-20 split.

- Training Set: 4,328 providers
- Testing Set: 1,082 providers

Stratified sampling was used to preserve the original fraud distribution in both datasets, ensuring representative model evaluation.

## Verify Class Distribution

In [23]:
print("Training Set Distribution")
print(y_train.value_counts())
print("\nTraining Set Percentage")
print(round(y_train.value_counts(normalize=True) * 100, 2))
print("\nTesting Set Distribution")
print(y_test.value_counts())
print("\nTesting Set Percentage")
print(round(y_test.value_counts(normalize=True) * 100, 2))

Training Set Distribution
Fraud
0    3923
1     405
Name: count, dtype: int64

Training Set Percentage
Fraud
0    90.64
1     9.36
Name: proportion, dtype: float64

Testing Set Distribution
Fraud
0    981
1    101
Name: count, dtype: int64

Testing Set Percentage
Fraud
0    90.67
1     9.33
Name: proportion, dtype: float64


The fraud distribution remains consistent across the training and testing datasets.

The use of stratified sampling successfully preserved the original class proportions, preventing sampling bias and ensuring reliable model evaluation.

## Standardize Numerical Features

In [24]:
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)
print("Scaling Completed Successfully")

Scaling Completed Successfully


Feature scaling was successfully applied using StandardScaler.

All numerical variables were transformed to have approximately zero mean and unit variance, which improves the performance of distance-based and linear machine learning algorithms.

## Verify Scaled Features

In [25]:
pd.DataFrame(X_train_scaled, columns=X.columns).head()

,InpatientClaimCount,UniqueInpatientBeneficiaries,AvgInpatientReimbursement,TotalInpatientReimbursement,AvgLengthOfStay,MaxLengthOfStay,AvgDiagnosisCount,OutpatientClaimCount,UniqueOutpatientBeneficiaries,AvgOutpatientReimbursement,TotalOutpatientReimbursement,AvgPatientAge,AvgChronicConditions,DeceasedRate,AvgIPAnnualReimbursement,AvgOPAnnualReimbursement,UniqueBeneficiaries
0,-0.314264,-0.322600,-0.665001,-0.309037,-0.677515,-0.614593,-0.785705,0.055817,0.252138,-0.033207,-0.010985,-0.135955,-0.591732,-0.250750,-0.413517,0.000033,0.198584
1,-0.314264,-0.322600,-0.665001,-0.309037,-0.677515,-0.614593,-0.785705,-0.301629,-0.393090,0.207926,-0.243953,-0.129077,0.503421,3.925048,-0.217392,0.940542,-0.422957
2,-0.314264,-0.322600,-0.665001,-0.309037,-0.677515,-0.614593,-0.785705,0.209008,0.384293,-0.024304,0.144098,-0.126054,0.092145,0.058568,-0.331092,-0.257280,0.325887
3,0.333501,0.264106,0.964189,0.296105,1.010262,0.553080,1.023255,-0.321269,-0.385317,0.147503,-0.295923,-1.312141,1.340104,-0.250750,0.322534,-0.514781,-0.325607
4,1.715399,1.877548,0.952562,1.573543,0.879321,1.428835,1.144067,-0.195574,-0.152102,-0.090805,-0.270696,0.771503,0.751079,-0.250750,0.813069,-0.395328,0.146165


### Interpretation

The transformed feature values are centered around zero and exhibit standardized ranges.

This confirms that scaling was correctly applied and that the dataset is ready for class balancing and machine learning model training.

## Handle Class Imbalance Using SMOTE

In [26]:
print("Before SMOTE")
print(y_train.value_counts())

Before SMOTE
Fraud
0    3923
1     405
Name: count, dtype: int64


### Interpretation

The training dataset exhibits significant class imbalance.

Fraudulent providers represent only a small proportion of observations, which may cause machine learning models to become biased toward the majority class.

In [27]:
smote = SMOTE(random_state=42)
X_train_smote, y_train_smote = smote.fit_resample(
    X_train_scaled,
    y_train
)
print("SMOTE Applied Successfully")

SMOTE Applied Successfully


## Verify Class Distribution After SMOTE

In [28]:
print("After SMOTE")
print(pd.Series(y_train_smote).value_counts())

After SMOTE
Fraud
0    3923
1    3923
Name: count, dtype: int64


### Interpretation

SMOTE generated synthetic fraud-provider observations to balance the training dataset.

The resulting dataset contains an equal number of fraudulent and non-fraudulent providers, enabling machine learning models to learn fraud patterns more effectively.

## Create Final Modeling Datasets

In [29]:
X_train_final = pd.DataFrame(X_train_smote, columns=X.columns)
X_test_final = pd.DataFrame(X_test_scaled, columns=X.columns)
y_train_final = pd.Series(y_train_smote, name="Fraud")
y_test_final = pd.Series(y_test, name="Fraud")
print("Final Modeling Datasets Created")

Final Modeling Datasets Created


The final modeling datasets have been successfully created.

Training data includes balanced fraud classes produced through SMOTE, while the testing dataset remains unchanged to provide an unbiased evaluation of model performance.

## Verify Final Dataset Dimensions

In [30]:
print("X_train_final:", X_train_final.shape)
print("X_test_final :", X_test_final.shape)
print("y_train_final:", y_train_final.shape)
print("y_test_final :", y_test_final.shape)

X_train_final: (7846, 17)
X_test_final : (1082, 17)
y_train_final: (7846,)
y_test_final : (1082,)


## Save Processed Modeling Datasets

In [31]:
import os
os.makedirs("../data/modeling", exist_ok=True)
X_train_final.to_csv("../data/modeling/X_train_final.csv", index=False)
X_test_final.to_csv("../data/modeling/X_test_final.csv", index=False)
y_train_final.to_csv("../data/modeling/y_train_final.csv", index=False)
y_test_final.to_csv("../data/modeling/y_test_final.csv", index=False)
print("Modeling Datasets Saved Successfully")

Modeling Datasets Saved Successfully


## Notebook Summary

This notebook prepared the provider-level healthcare fraud dataset for machine learning model development.

### Completed Steps

#### Data Preparation
- Loaded the provider-level feature dataset.
- Verified dataset structure, dimensions, and data types.
- Confirmed the absence of missing values.

#### Feature Selection
- Removed identifier and target-related columns.
- Selected 17 provider-level predictor variables.
- Created the feature matrix (X) and target vector (y).

#### Train-Test Split
- Split the dataset into training and testing subsets using an 80-20 ratio.
- Applied stratified sampling to preserve the original fraud distribution.

#### Feature Scaling
- Standardized all numerical features using StandardScaler.
- Transformed features to comparable scales for machine learning algorithms.

#### Class Imbalance Handling
- Identified significant class imbalance in the training dataset.
- Applied SMOTE (Synthetic Minority Oversampling Technique) to generate synthetic fraud-provider observations.
- Created a balanced training dataset with equal representation of fraud and non-fraud providers.

#### Final Modeling Dataset Creation
- Generated final training and testing datasets.
- Preserved the original testing dataset for unbiased model evaluation.

### Output Files

The following datasets were saved for machine learning modeling:

- X_train_final.csv
- X_test_final.csv
- y_train_final.csv
- y_test_final.csv

### Outcome

The healthcare provider dataset is now fully prepared for machine learning model development. The processed datasets will be used in the next notebook to train, evaluate, and compare multiple fraud detection algorithms including Logistic Regression, Random Forest, and XGBoost.